microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/services/properties/PropertiesFileInlayHint.java

```diff
@@ -0,0 +1,97 @@
+*
+* SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+*
+* Contributors:
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package org.eclipse.lsp4mp.services.properties;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.eclipse.lsp4j.InlayHint;
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.Range;
+import org.eclipse.lsp4j.jsonrpc.CancelChecker;
+import org.eclipse.lsp4mp.commons.MicroProfileProjectInfo;
+import org.eclipse.lsp4mp.ls.commons.BadLocationException;
+import org.eclipse.lsp4mp.model.Node;
+import org.eclipse.lsp4mp.model.Node.NodeType;
+import org.eclipse.lsp4mp.model.PropertiesModel;
+import org.eclipse.lsp4mp.model.Property;
+import org.eclipse.lsp4mp.model.PropertyGraph;
+import org.eclipse.lsp4mp.model.PropertyValue;
+
+/**
+ * The properties file inlay hint support.
+ * 
+ * Given this properties file:
+ * 
+ * <code>
+    server.url=https://${host}:${port:8080}/${endpoint}
+	host=microprofile.io
+	app=project
+	service=eclipse/microprofile-config
+	endpoint=${app}/${service}
+ * </code>
+ * 
+ * Inlay hint will be displayed for the 2 properties which have expression:
+ * 
+ * <code>
+    server.url=https://${host}:${port:8080}/${endpoint} [https://microprofile.io:8080/project/eclipse/microprofile-config]
+	host=microprofile.io
+	app=project
+	service=eclipse/microprofile-config
+	endpoint=${app}/${service} [eclipse/microprofile-config]
+ * </code>
+ */
+
+class PropertiesFileInlayHint {
+
+	private static final Logger LOGGER = Logger.getLogger(PropertiesFileInlayHint.class.getName());
+
+	public List<InlayHint> getInlayHint(PropertiesModel document, MicroProfileProjectInfo projectInfo, Range range,
+			CancelChecker cancelChecker) {
+		List<InlayHint> hints = new ArrayList<>();
+		List<Node> children = document.getChildren();
+		for (Node child : children) {
+			cancelChecker.checkCanceled();
+			if (child.getNodeType() == NodeType.PROPERTY) {
+				Property property = (Property) child;
+				PropertyValue valueNode = property.getValue();
+				if (valueNode != null && valueNode.hasExpression()) {
+					// The current property has a value with expression:
+					// ex : server.url=https://${host}:${port:8080}/${endpoint}
+					PropertyGraph graph = new PropertyGraph(document);
+					String resolved = valueNode.getResolvedValue(graph, projectInfo, cancelChecker);
+					if (resolved != null) {
+						try {
+							// The expression 'https://${host}:${port:8080}/${endpoint}' can be resolved
+							// ex : https://microprofile.io:8080/project/eclipse/microprofile-config
+							// Display this resolved with inlay hint:
+							// server.url=https://${host}:${port:8080}/${endpoint}
+							// [https://microprofile.io:8080/project/eclipse/microprofile-config]
+							InlayHint hint = new InlayHint();
+							hint.setLabel(" " + resolved);
```

Comment:
I think it might be easier to notice the inlay hint if there were multiple spaces between the end of the line and the inlay hint, for instance like [this extension](https://marketplace.visualstudio.com/items?itemName=usernamehw.errorlens) does it. What do you think?

Reply:
Sorry I don't understand what you mean? You mean tomanage some coloration? I'm not sure that it is possible with LSP.

Reply:
Not the colouration, just to add more whitespace. eg. here (`[` `]` is to show where the inlay hint is)

now
```properties
my.value=${asdf} asdf [ 9090 asdf]
asdf=9090
```
vs suggestion:
```properties
my.value=${asdf} asdf [    9090 asdf]
asdf=9090
```
I don't know if it will look better, since it looks like the background where there are spaces will be a different colour, which could be confusing.

Reply:
Yeah, my suggestion looks a little odd with the default look of inlay hints:

![InlayHintAdditionalSpaces](https://user-images.githubusercontent.com/22376627/186939390-73a7f920-8d9d-443f-9aac-7012a494fe53.png)